### ISSUE TITLE
Consider carry custom extension fields in svid

### ISSUE THREAD

**Issue by chenxi-neo** [spiffe/spiffe#254]
We hope to be able to carry custom extension fields in jwt & X509 svids to meet business needs. Currently, we use spiffe/spire as the basic authentication infrastructure for micro-service. Some of our internal upstream businesses hope to carry some custom extension fields through jwt or x509, such as: trace_id, taskname, etc. This information is parsed in the downstream business for authentication or other purpose . Both JWT and X509 standards support custom extension fields. JWT can be based on custom claims, and X509 can be implemented through v3 extension. We would like to know how the community thinks about this proposal, and whether this specification can be written into the spiffe standard and implemented in spire. Currently, we are exploring this proposal internally, and if the community is interested, we can also work with the community to build this capability. If the community does not accept the proposal, does the community have a better idea of the above problems?

**Comment by spikecurtis** [comment]
Thanks for opening and issue!

I have a few questions to try to understand your request in some more detail.

As you mention, JWT and x509 support custom extension fields.  The SPIFFE specs for these SVID types are written such that they allow custom extensions to be included.  E.g. from the JWT-SVID spec:

> The JWT-SVID specification does not introduce any new claims, though it does set some restrictions on the registered claims defined by [RFC 7519](https://tools.ietf.org/html/rfc7519). Registered claims not described in this document, in addition to private claims, MAY be used as implementers see fit.

So, is your request that SPIFFE specs allow custom claims in SVIDs?  If so, I believe that is the current state of affairs.

You mention implementation in SPIRE: is the request that SPIRE provide an API to allow you to inject custom claims into SVIDs it mints?  If so, I think this request would be better directed to https://github.com/spiffe/spire.  The spec working group standardizes the Workload-API that SPIRE implements, but does not specify the administrative or operator APIs that SPIRE uses.

Or, is your request not covered by either of the above situations and I'm just not understanding what you are asking.  If so, would you mind restating your proposal with some more detail?

**Comment by chenxi-neo** [comment]
Thank you for your response.
It seems that there's no restrictions for spiffe to add custom jwt svid cliams as you mentioned above. And how about X509? seems there's also no further restriction about that?
If there are no restrictions about customizing extension of both jwt and x509 svid, we will also discuss with spire teams to support this feature.

**Comment by azdagron** [comment]
I will point out that SPIRE supports a [CredentialComposer](https://github.com/spiffe/spire-plugin-sdk/blob/main/proto/spire/plugin/server/credentialcomposer/v1/credentialcomposer.proto) plugin that operators can implement to add arbitrary extensions to both X.509 and JWT SVIDs. This was added somewhat recently.

**Comment by chenxi-neo** [comment]
> I will point out that SPIRE supports a [CredentialComposer](https://github.com/spiffe/spire-plugin-sdk/blob/main/proto/spire/plugin/server/credentialcomposer/v1/credentialcomposer.proto) plugin that operators can implement to add arbitrary extensions to both X.509 and JWT SVIDs. This was added somewhat recently.

It looks very cool. And do you have any design specification and using guideline for developers to dive deep into?

**Comment by chenxi-neo** [comment]
@azdagron i quickly look through the source code about the CredentialComposer, i agree that developers could leverage this plugin to add extensions to X.509 and JWT SVIDs. However, if the extension values are binded to workload itself (eg. trace id of the task which is running on certain workload), how could we deal with this problem?

**Comment by spikecurtis** [comment]
> It seems that there's no restrictions for spiffe to add custom jwt svid cliams as you mentioned above. And how about X509? seems there's also no further restriction about that?

That's right.  Although maybe the spec could be more explicit on this point, the SVID spec is meant to be interpreted as restricting a few key fields/extensions of an otherwise valid x509 certificate.  Any fields/extensions not mentioned in the SPIFFE specs are unrestricted.

**Comment by azdagron** [comment]
Hey @chenxi-neo, at this point we're diving into the weeds with SPIRE. Do you mind if I transfer this issue over to that repo?

**Comment by chenxi-neo** [comment]
> Hey @chenxi-neo, at this point we're diving into the weeds with SPIRE. Do you mind if I transfer this issue over to that repo?

i think it's a good idea to transfer this discussion to spire repo.
